README.md
# Pi Dashboard Go
**Pi Dashboard Go** is a Golang implementation of pi-dashboard

* **[中文文档](https://ojoll.com/archives/86/)**



![](./screenshots/screenshot_index.png)

![](./screenshots/screenshot_login.png)

![](./screenshots/screenshot_index_dark.png)

![](./screenshots/screenshot_login_dark.png)





## Install

Thanks to the characteristics of the **[Golang](https://golang.org/)** language, the deployment of **Pi Dashboard Go** is very simple: **single binary executable file**.

#### Download

Just download the executable file from the project **[Releases](https://github.com/plutobell/pi-dashboard-go/releases)** page, **no other dependencies**.

#### Authority

Grant executable permissions

```
chmod +x pi-dashboard-go
```

**Note：Pi Dashboard Go requires root privileges.**



## Use

#### Usage

**Pi Dashboard Go** can be configured via command line parameters：

```bash
Pi Dashboard Go version: v1.7.0
Project address: https://github.com/plutobell/pi-dashboard-go

Usage: Pi Dashboard Go [-auth USR:PSW] [-disk Paths] [-help]
[-interval Seconds] [-log] [-net NIC] [-port Port]
[-session Days] [-theme Theme] [-title Title] [-version]

Options:
  -auth string
        specify username and password (default "pi:123")
  -disk string
        specify the filesystem path (default "/")
  -help
        this help
  -interval string
        specify the update interval in seconds (default "1")
  -log
        enable log display
  -net string
        specify the network device (default "lo")
  -port string
        specify the running port (default "8080")
  -session string
        specify the login status validity in days (default "7")
  -theme string
        specify the theme between 'light' and 'dark' (default "light")
  -title string
        specify the website title (default "Pi Dashboard Go")
  -version
        show version and exit
```



## Thanks

* **[Pi Dashboard](https://github.com/spoonysonny/pi-dashboard)**
* **[echo](https://github.com/labstack/echo)**
* **[gopsutil](https://github.com/shirou/gopsutil)**

* **[bootstrap](https://github.com/twbs/bootstrap)**
* **[jquery](https://github.com/jquery/jquery)**
* **[highcharts](https://github.com/highcharts/highcharts)**

## Changelog

* **[Changelog](./CHANGELOG.md)**

CHANGELOG.md
# Changelog #

**2023-04-05**

* v1.7.0 : 
  * Added a cli parameter called theme #4
  * Added proxy for version detection
  * Fixed the bug of disk path detection
  * Some other optimizations
  * Updated dependencies
  * Use Go v1.20.3

**2021-09-02**

* v1.6.0 : 
  * Refactored device package to significantly improve performance and speed
  * Improved front-end display details
  * Updated dependencies

**2021-08-28**

* v1.5.1 : 
  * Fixed the bug of load average display
  * Added the navbar to close automatically when clicked on mobile devices

**2021-08-27**

* v1.5.0 : 
  * Added dark mode that follows the device switch
  * Added automatic jumping to the login page when not logged in
  * Some other optimizations
  * Updated dependencies

**2021-08-24**

* v1.4.1 : 
  * Added operation functions permission detection
  * Added a prompt to the console for non-root users to run
* v1.4.0 : 
  * Adjusted the front-end page display style
  * Adjusted static file path structure
  * Added page loading time statistics
  * Added pop-up window for new version release notes display
  * Some other optimizations
  * Use Go v1.17
  * Updated dependencies

**2021-08-14**

* v1.3.3 : 
  * Added automatic recognition and switching of favicon
  * Adjusted static file path structure
  * Adjusted project structure
  * Updated dependencies

**2021-08-13**

* v1.3.2 : 
  * Added automatic version detection and prompting
* v1.3.1 : 
  * Added csrf protection
  * Adjusted some details
  * Updated dependencies

**2021-08-12**

* v1.3.0 : 
  * Adjusted routing structure
  * Adjusted log formatting
  * Added gzip compression
  * Enhanced web security
  * Login request changed to asynchronous
  * Refactored part of the code
  * Updated dependencies

**2021-08-10**

* v1.2.1 : 
  * Added  login empty field checks
  * Added login failure prompt message
  * Added new command line parameter: log

* v1.2.0 : 
  * Rewrite login authentication
  * Added new login page
  * Added new command line parameter: session
  * Use Go v1.16.7
  * Updated dependencies

**2021-06-17**

* v1.1.2 : 
  * Changed the way cpu usage is calculated
  * Updated dependencies

**2021-06-16**

* v1.1.1 : 
  * Fix the bug of abnormal display of cpu information
  * Use Go v1.16.5
  * Updated dependencies

**2021-04-05**

* v1.1.0 : 
  * Replace go.rice with go embed
  * Fix the bug of program error when unable to get cpu model
  * Added new command line parameter: interval
  * Added a different header image for non-Raspberry Pi devices

**2021-03-31**

* v1.0.10 :
  * Fix the bug of panic caused by empty device model information #1
  * Fix the bug of not finding dashboard.min.js #2
  * Fix the bug of invalid hostname command on arch
  * Added golang version display on view

**2020-8-14**

* v1.0.9 : 
  * Fix swap display bug
  * Adapt to linux system under 386 and amd64
  * Optimize error handling for function Popen

**2020-8-9**

* v1.0.8 : 
  * Optimize swap display details
  * Added shortcut buttons such as shutdown and reboot

**2020-8-7**

* v1.0.7 : 
  * Optimize network card flow and curve display
  * Interface detail adjustment

**2020-8-6**

* v1.0.6 : 
  * Fix the bug that the network card data display error
  * Fixed navigation bar at the top
  * Interface detail adjustment
* v1.0.5 : 
  * Interface color adjustment
  * Data update detection and prompt
  * Optimize code for server
  * Detail adjustment
* v1.0.4 : 
  * Adjust Cached calculation method
  * Added theme-color for mobile browser
  * Added display login user statistics
  * Bug fixes and details optimization

**2020-8-5**

* v1.0.3 : 
  * Newly added time formatting function resolveTime
  * Detail optimization
* v1.0.2 : 
  * Improve command line parameter verification
  * Detail optimization
  * Added test case device_test.go
  * New page loading animation

**2020-8-4**

* v1.0.1 : Bug fixes, detail optimization
* v1.0.0

main.go
// @Program : Pi Dashboard Go (https://github.com/plutobell/pi-dashboard-go)
// @Description: Golang implementation of pi-dashboard
// @Author: github.com/plutobell
// @Creation: 2020-08-01
// @Last modification: 2023-04-05
// @Version: 1.7.0

package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"os/user"
	"path/filepath"
	"strconv"
	"strings"
	"unicode"

	"github.com/plutobell/pi-dashboard-go/config"
	"github.com/plutobell/pi-dashboard-go/device"
	"github.com/plutobell/pi-dashboard-go/server"
)

func init() {
	flag.BoolVar(&config.Help, "help", false, "this help")
	flag.BoolVar(&config.Version, "version", false, "show version and exit")
	flag.StringVar(&config.Port, "port", "8080", "specify the running port")
	flag.StringVar(&config.Title, "title", "Pi Dashboard Go", "specify the website title")
	flag.StringVar(&config.Net, "net", "lo", "specify the network device")
	flag.StringVar(&config.Disk, "disk", "/", "specify the filesystem path")
	flag.StringVar(&config.Auth, "auth", config.USERNAME+":"+config.PASSWORD, "specify username and password")
	flag.StringVar(&config.Interval, "interval", "1", "specify the update interval in seconds")
	flag.StringVar(&config.SessionMaxAge, "session", "7", "specify the login status validity in days")
	flag.StringVar(&config.Theme, "theme", "light", "specify the theme between 'light' and 'dark'")
	flag.BoolVar(&config.EnableLogger, "log", false, "enable log display")

	config.SessionName = "logged_in"
	config.FileName = filepath.Base(os.Args[0])
	config.LinuxUserInfo, _ = user.Current()

	flag.Usage = usage
}

func main() {
	flag.Parse()

	if config.Help {
		flag.Usage()
		return
	}
	if config.Version {
		fmt.Println("Pi Dashboard Go v" + config.VERSION)
		fmt.Println("Project address: " + config.PROJECT)
		return
	}
	netDevs, err := device.Popen("cat /proc/net/dev")
	if err != nil {
		log.Fatal(err)
		return
	}
	if !strings.Contains(netDevs, config.Net+":") {
		fmt.Println("Network card does not exist")
		return
	}
	pathExists := false
	_, err = os.Stat(config.Disk)
	if err == nil {
		pathExists = true
	}
	if os.IsNotExist(err) {
		pathExists = false
	}

	if config.Disk != "/" {
		if !pathExists {
			fmt.Println("Disk does not exist")
			return
		}
	}
	authSlice := strings.Split(config.Auth, ":")
	if len(authSlice) != 2 {
		fmt.Println("Auth format error")
		return
	}
	if len([]rune(authSlice[0])) > 15 || len([]rune(authSlice[0])) == 0 {
		fmt.Println("Username is too long")
		return
	}
	if len([]rune(authSlice[1])) > 15 || len([]rune(authSlice[1])) == 0 {
		fmt.Println("Password is too long")
		return
	}
	if len([]rune(config.Title)) > 25 {
		fmt.Println("Title is too long")
		return
	}

	isDigit := true
	for _, r := range config.Interval {
		if !unicode.IsDigit(rune(r)) {
			isDigit = false
			break
		}
	}
	if !isDigit {
		fmt.Println("Interval parameter value is invalid")
		return
	}

	IntervalInt, err := strconv.Atoi(config.Interval)
	if err != nil {
		log.Fatal(err)
		return
	}
	if IntervalInt > 900 {
		fmt.Println("Interval is too long")
		return
	} else if IntervalInt < 0 {
		fmt.Println("Interval should be no less than 0")
		return
	}

	SessionMaxAgeInt, err := strconv.Atoi(config.SessionMaxAge)
	if err != nil {
		log.Fatal(err)
		return
	}
	if SessionMaxAgeInt > 365 {
		fmt.Println("Session days is too long")
		return
	} else if SessionMaxAgeInt < 0 {
		fmt.Println("Session days should be no less than 0")
		return
	}

	if config.Theme != "light" && config.Theme != "dark" {
		fmt.Println("Theme name not supported")
		return
	}

	server.Run()
}

func usage() {
	fmt.Fprintf(os.Stderr, `Pi Dashboard Go version: v%s
Project address: %s

Usage: %s [-auth USR:PSW] [-disk Paths] [-help]
[-interval Seconds] [-log] [-net NIC] [-port Port]
[-session Days] [-theme Theme] [-title Title] [-version]

Options:
`, config.VERSION, config.PROJECT, config.FileName)
	flag.PrintDefaults()
}
